editor/src/components/inspector/common/name-and-icon-hook.ts

```diff
@@ -53,6 +50,19 @@ export function useNameAndIcon(path: TemplatePath): NameAndIconResult {
   )
 }
 
+export function useNamesAndIconsSelectedViews(): NameAndIconResult[] {
+  return useEditorState(
+    (store) => {
+      const metadata = store.editor.jsxMetadataKILLME
+      const components = getOpenUtopiaJSXComponentsFromState(store.editor)
+      const imports = getOpenImportsFromState(store.editor)
+      const selectedViews = store.editor.selectedViews
+      return selectedViews.map(path => getNameAndIconResult(path, components, metadata, imports))
```

Comment:
by default `useEditorState` uses a shallow equality comparison. so if you return an array of stuff, it will use `===` comparison for each array item. 

however `getNameAndIconResult` will always return a new object, so even if you call it with the same parameters,
it will never satisfy `===` equality, and it will always trigger a re-render of the component using this hook, for every single editor state change.

One trick is while developing a new component or new hook, I recommend turning on React Devtool's "Highlight updates when components render" feature.

![](https://miro.medium.com/max/1874/1*R0ij8ll0ZiPxwS7Q1xDUKg.png)

That will flash your component whenever it re-renders. If you see your component flashing while you are scrolling the canvas for example, that means that your component is re-rendering too often. _Ideally_ it should only flash when it actually changes what is displayed. It is sometimes too complicated to achieve perfection, so you can use deliberation and decide how much optimization is necessary. But we always want to avoid a component which re-renders on every change!

There's two potential solutions:
one is to use memoization on this selector (the function passed into useEditorState is called a selector). There's some selector memoization helpers called `reselect` that we use in some places in the codebase: https://github.com/concrete-utopia/utopia/blob/d4c971fc0587cda5c880220a270e09e0191172ee/editor/src/components/canvas/controls/new-canvas-controls.tsx#L211-L248

The second (less good, but sometimes necessary) option is to roll your own equality function. You can provide a third (optional) parameter to `useEditorState` which is an equality function in the form of `(oldSlice, newSlice) => boolean`, if this equality function returns true, we consider the old state slice and the new state slice to be unchanged, and we do not trigger a re-render of the element. Here's an example for that: https://github.com/concrete-utopia/utopia/blob/d4c971fc0587cda5c880220a270e09e0191172ee/editor/src/components/inspector/common/name-and-icon-hook.ts#L34-L54

Reply:
Todo!

Reply:
Extremely helpful comment Balazs, thanks for taking the time to write this out!

Reply:
fixed this issue, so the PR can be merged today